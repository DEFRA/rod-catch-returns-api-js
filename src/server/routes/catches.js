import {
  mapCatchToResponse,
  mapRequestToCatch
} from '../../mappers/catches.mapper.js'
import { Catch } from '../../entities/index.js'
import { StatusCodes } from 'http-status-codes'
import logger from '../../utils/logger-utils.js'

export default [
  {
    method: 'POST',
    path: '/catches',
    options: {
      /**
       * Create a catch (salmon and large sea trout) for an activity in the database
       *
       * @param {import('@hapi/hapi').Request request - The Hapi request object
       *     @param {string} request.payload.activity - The activity id prefixed with activities/
       *     @param {string} request.payload.dateCaught - The ISO 8601 date the catch was recorded (e.g., "2024-06-24T00:00:00+01:00").
       *     @param {string} request.payload.species - The species ID prefixed with "species/".
       *     @param {Object} request.payload.mass - The mass of the catch.
       *     @param {number} request.payload.mass.kg - The mass in kilograms.
       *     @param {number} request.payload.mass.oz - The mass in ounces.
       *     @param {string} request.payload.mass.type - The mass type, typically "IMPERIAL" or "METRIC".
       *     @param {string} request.payload.method - The method ID prefixed with "methods/".
       *     @param {boolean} request.payload.released - Indicates if the catch was released.
       *     @param {boolean} request.payload.onlyMonthRecorded - Indicates if only the month was recorded, to allow FMT users to report on the default dates
       *     @param {boolean} request.payload.noDateRecorded - Indicates if no specific date was recorded, to allow FMT users to report on the default month
       *     @param {string} request.payload.reportingExclude - Is this entry excluded from reporting
       * @param {import('@hapi/hapi').ResponseToolkit} h - The Hapi response toolkit
       * @returns {Promise<import('@hapi/hapi').ResponseObject>} - A response containing the target {@link Activity}
       */
      handler: async (request, h) => {
        try {
          const catchData = mapRequestToCatch(request.payload)

          const createdCatch = await Catch.create(catchData)

          const catchResponse = mapCatchToResponse(
            request,
            createdCatch.toJSON()
          )

          return h.response(catchResponse).code(StatusCodes.CREATED)
        } catch (error) {
          logger.error('Error create catch:', error)
          return h
            .response({ error: 'Unable to create catch' })
            .code(StatusCodes.INTERNAL_SERVER_ERROR)
        }
      },
      description:
        'Create a catch (salmon and large sea trout) for an activity in the database',
      notes:
        'Create a catch (salmon and large sea trout) for an activity in the database',
      tags: ['api', 'catches']
    }
  }
]
