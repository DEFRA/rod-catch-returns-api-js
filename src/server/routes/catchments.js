import { Catchment } from '../../entities/catchment.entity.js'
import { StatusCodes } from 'http-status-codes'
import logger from '../../utils/logger-utils.js'

export default [
  {
    method: 'GET',
    path: '/catchments',
    options: {
      /**
       * Retrieve all the catchments in the database
       *
       * @param {import('@hapi/hapi').Request request - The Hapi request object
       * @param {import('@hapi/hapi').ResponseToolkit} h - The Hapi response toolkit
       * @returns {Promise<import('@hapi/hapi').ResponseObject>} - A response containing the target {@link Catchment}
       */
      handler: async (_request, h) => {
        try {
          const catchments = await Catchment.findAll()
          return h
            .response({
              _embedded: {
                catchments
              }
            })
            .code(StatusCodes.OK)
        } catch (error) {
          logger.error('Error fetching catchments:', error)
          return h
            .response({ error: 'Unable to fetch catchments' })
            .code(StatusCodes.INTERNAL_SERVER_ERROR)
        }
      },
      description: 'Retrieve all the catchments in the database',
      notes: 'Retrieve all the catchments in the database',
      tags: ['api', 'catchments']
    }
  }
]
