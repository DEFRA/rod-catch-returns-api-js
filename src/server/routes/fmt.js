import { authorize, signIn } from '../../services/oidc-service.js'

export default [
  {
    path: '/oidc/authorize',
    method: 'GET',
    handler: authorize,
    options: { auth: false }
  },
  {
    path: '/oidc/callback',
    method: 'POST',
    handler: signIn,
    options: { auth: false }
  }
]
