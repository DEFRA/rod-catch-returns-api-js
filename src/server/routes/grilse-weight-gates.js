import { GrilseWeightGate } from '../../entities/index.js'
import { StatusCodes } from 'http-status-codes'
import { handleServerError } from '../../utils/server-utils.js'
import { mapGrilseWeightGateToResponse } from '../../mappers/grilse-weight-gates.mapper.js'

export default [
  {
    method: 'GET',
    path: '/grilseWeightGates',
    options: {
      /**
       * Retrieve all the grilse weight gates in the database
       *
       * @param {import('@hapi/hapi').Request request - The Hapi request object
       * @param {import('@hapi/hapi').ResponseToolkit} h - The Hapi response toolkit
       * @returns {Promise<import('@hapi/hapi').ResponseObject>} - A response containing the target {@link GrilseWeightGate}
       */
      handler: async (request, h) => {
        try {
          const grilseWeightGates = await GrilseWeightGate.findAll()

          const mappedGrilseWeightGates = grilseWeightGates.map(
            (grilseWeightGate) =>
              mapGrilseWeightGateToResponse(grilseWeightGate)
          )

          return h
            .response({
              _embedded: {
                grilseWeightGates: mappedGrilseWeightGates
              }
            })
            .code(StatusCodes.OK)
        } catch (error) {
          return handleServerError(
            'Error fetching grilse weight gates',
            error,
            h
          )
        }
      },
      description: 'Retrieve all the grilse weight gates in the database',
      notes: 'Retrieve all the grilse weight gates in the database',
      tags: ['api', 'grilseWeightGates']
    }
  }
]
