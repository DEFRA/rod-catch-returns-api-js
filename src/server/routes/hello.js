export default [
  {
    method: 'GET',
    path: '/hello',
    /**
     * Handler that return hello.
     *
     * @param {import('@hapi/hapi').Request request - The Hapi request object
     * @param {import('@hapi/hapi').ResponseToolkit} h - The Hapi response toolkit
     * @returns {import('@hapi/hapi').ResponseObject} - A response containing hello
     */
    handler: async (request, h) => {
      return h.response('Hello mate').code(200)
    }
  }
]
