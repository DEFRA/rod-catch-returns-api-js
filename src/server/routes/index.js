import Catchments from './catchments.js'
import Licences from './licences.js'
import Region from './regions.js'
import Rivers from './rivers.js'
import Species from './species.js'

import Static from './static.js'

export const apiPrefixRoutes = [].concat(Licences, Region, Rivers, Catchments, Species)

export const rootRoutes = [].concat(Static)
