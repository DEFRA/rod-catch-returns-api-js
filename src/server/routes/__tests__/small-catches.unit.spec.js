import { SmallCatch, SmallCatchCount } from '../../../entities/index.js'
import {
  getMockResponseToolkit,
  getServerDetails
} from '../../../test-utils/server-test-utils.js'
import {
  handleNotFound,
  handleServerError
} from '../../../utils/server-utils.js'
import routes from '../small-catches.js'
import { sequelize } from '../../../services/database.service.js'

jest.mock('../../../entities/index.js')
jest.mock('../../../utils/logger-utils.js')
jest.mock('../../../utils/server-utils.js')
jest.mock('../../../services/database.service.js', () => ({
  sequelize: {
    transaction: jest.fn(),
    define: jest.fn(() => ({
      associate: jest.fn(),
      bulkCreate: jest.fn(),
      hasMany: jest.fn(),
      belongsTo: jest.fn(),
      findAll: jest.fn(),
      findOne: jest.fn(),
      create: jest.fn(),
      destroy: jest.fn(),
      associations: jest.fn()
    })),
    literal: jest.fn()
  }
}))

const [
  {
    options: { handler: postSmallCatchHandler }
  },
  {
    options: { handler: getActivityForSmallCatchHandler }
  },
  {
    options: { handler: getSmallCatchHandler }
  },
  {
    options: { handler: deleteSmallCatchHandler }
  },
  {
    options: { handler: patchSmallCatchHandler }
  }
] = routes

const NOT_FOUND_SYMBOL = Symbol('NOT_FOUND')
const SERVER_ERROR_SYMBOL = Symbol('SERVER_ERROR')

handleNotFound.mockReturnValue(NOT_FOUND_SYMBOL)
handleServerError.mockReturnValue(SERVER_ERROR_SYMBOL)

describe('smallCatches.unit', () => {
  describe('POST /smallCatches', () => {
    const getSmallCatchRequest = () => ({
      ...getServerDetails(),
      payload: {
        submission: 'submissions/2802',
        activity: 'activities/3',
        month: 'FEBRUARY',
        released: '3',
        counts: [
          { method: 'methods/1', count: '3' },
          { method: 'methods/2', count: '2' },
          { method: 'methods/3', count: '1' }
        ],
        noMonthRecorded: false,
        reportingExclude: false
      }
    })

    const getCreatedSmallCatch = () => ({
      toJSON: jest.fn().mockReturnValue({
        id: '1',
        month: 2,
        released: 3,
        counts: [
          {
            count: 3,
            method_id: 1
          },
          {
            count: 2,
            method_id: 2
          },
          {
            count: 1,
            method_id: 3
          }
        ],
        activity_id: 3,
        noMonthRecorded: false,
        reportingExclude: false,
        createdAt: '2024-10-11T09:30:57.463+0000',
        updatedAt: '2024-10-11T09:30:57.463+0000',
        version: '2024-10-11T09:30:57.463+0000'
      })
    })

    afterEach(() => {
      jest.clearAllMocks()
    })

    it('should return a 201 status code if the small catch is created successfully', async () => {
      SmallCatch.create.mockResolvedValueOnce(getCreatedSmallCatch())

      const result = await postSmallCatchHandler(
        getSmallCatchRequest(),
        getMockResponseToolkit()
      )

      expect(result.statusCode).toBe(201)
    })

    it('should return the created small catch if the call to create it is successful', async () => {
      SmallCatch.create.mockResolvedValueOnce(getCreatedSmallCatch())

      const result = await postSmallCatchHandler(
        getSmallCatchRequest(),
        getMockResponseToolkit()
      )

      expect(result.payload).toMatchSnapshot()
    })

    it('should call handleServerError if an error occurs while creating the small catch', async () => {
      const error = new Error('Database error')
      SmallCatch.create.mockRejectedValueOnce(error)
      const h = getMockResponseToolkit()

      await postSmallCatchHandler(getSmallCatchRequest(), h)

      expect(handleServerError).toHaveBeenCalledWith(
        'Error creating small catch',
        error,
        h
      )
    })

    it('should return an error response if an error occurs while creating the small catch', async () => {
      const error = new Error('Database error')
      SmallCatch.create.mockRejectedValueOnce(error)

      const result = await postSmallCatchHandler(
        getSmallCatchRequest(),
        getMockResponseToolkit()
      )

      expect(result).toBe(SERVER_ERROR_SYMBOL)
    })
  })

  describe('GET /smallCatches/{smallCatchId}/activity', () => {
    const getSmallCatchRequest = (smallCatchId) =>
      getServerDetails({
        params: {
          smallCatchId
        }
      })

    const getSmallCatchWithActivity = () => ({
      toJSON: jest.fn().mockReturnValue({
        id: '1',
        month: 1,
        released: 1,
        reportingExclude: false,
        noMonthRecorded: false,
        createdAt: '2024-10-10T13:27:41.096Z',
        updatedAt: '2024-10-10T13:27:41.096Z',
        version: '2024-10-10T13:27:41.097Z',
        ActivityId: '3',
        activity_id: '3',
        Activity: {
          id: '3',
          daysFishedWithMandatoryRelease: 1,
          daysFishedOther: 0,
          createdAt: '2024-10-10T13:26:30.344Z',
          updatedAt: '2024-10-10T13:26:30.344Z',
          version: '2024-10-10T13:26:30.345Z',
          submission_id: '2802',
          river_id: '104',
          SubmissionId: '2802'
        }
      })
    })

    afterEach(() => {
      jest.clearAllMocks()
    })

    it('should return a 200 status code and the activity if the small catch and activity is found', async () => {
      SmallCatch.findOne.mockResolvedValueOnce(getSmallCatchWithActivity())

      const result = await getActivityForSmallCatchHandler(
        getSmallCatchRequest('1'),
        getMockResponseToolkit()
      )

      expect(result.payload).toMatchSnapshot()
      expect(result.statusCode).toBe(200)
    })

    it('should call handleNotFound if the activity for the small catch is not found', async () => {
      SmallCatch.findOne.mockResolvedValueOnce(null)
      const h = getMockResponseToolkit()

      await getActivityForSmallCatchHandler(
        getSmallCatchRequest('nonexistent-id'),
        h
      )

      expect(handleNotFound).toHaveBeenCalledWith(
        'Activity not found for small catch with ID nonexistent-id',
        h
      )
    })

    it('should return a not found response if the activity for the small catch is not found', async () => {
      SmallCatch.findOne.mockResolvedValueOnce(null)
      const h = getMockResponseToolkit()

      const result = await getActivityForSmallCatchHandler(
        getSmallCatchRequest('nonexistent-id'),
        h
      )

      expect(result).toBe(NOT_FOUND_SYMBOL)
    })

    it('should call handleServerError if an error occurs while fetching the activity for the small catch', async () => {
      const error = new Error('Database error')
      SmallCatch.findOne.mockRejectedValueOnce(error)
      const h = getMockResponseToolkit()

      await getActivityForSmallCatchHandler(getSmallCatchRequest('1'), h)

      expect(handleServerError).toHaveBeenCalledWith(
        'Error fetching activity for small catch',
        error,
        h
      )
    })

    it('should an error response if an error occurs while fetching the activity for the small catch', async () => {
      const error = new Error('Database error')
      SmallCatch.findOne.mockRejectedValueOnce(error)
      const h = getMockResponseToolkit()

      const result = await getActivityForSmallCatchHandler(
        getSmallCatchRequest('1'),
        h
      )

      expect(result).toBe(SERVER_ERROR_SYMBOL)
    })
  })

  describe('GET /smallCatches/{smallCatchId}', () => {
    const getSmallCatchRequest = (smallCatchId) =>
      getServerDetails({
        params: {
          smallCatchId
        }
      })

    const getSmallCatch = () => ({
      toJSON: jest.fn().mockReturnValue({
        id: '33551',
        month: 1,
        released: 1,
        reportingExclude: false,
        noMonthRecorded: false,
        createdAt: '2024-12-11T10:07:11.726Z',
        updatedAt: '2024-12-11T10:07:11.726Z',
        version: '2024-12-11T10:07:11.729Z',
        ActivityId: '133750',
        activity_id: '133750',
        counts: [
          {
            small_catch_id: '33551',
            method_id: '1',
            count: 2,
            SmallCatchId: '33551'
          },
          {
            small_catch_id: '33551',
            method_id: '3',
            count: 2,
            SmallCatchId: '33551'
          }
        ]
      })
    })

    afterEach(() => {
      jest.clearAllMocks()
    })

    it('should return a 200 status code and the small catch if it is found', async () => {
      SmallCatch.findOne.mockResolvedValueOnce(getSmallCatch())

      const result = await getSmallCatchHandler(
        getSmallCatchRequest('1'),
        getMockResponseToolkit()
      )

      expect(result.payload).toMatchSnapshot()
      expect(result.statusCode).toBe(200)
    })

    it('should call handleNotFound if the small catch is not found', async () => {
      SmallCatch.findOne.mockResolvedValueOnce(null)
      const h = getMockResponseToolkit()

      await getSmallCatchHandler(getSmallCatchRequest('nonexistent-id'), h)

      expect(handleNotFound).toHaveBeenCalledWith(
        'Small catch not found for ID: nonexistent-id',
        h
      )
    })

    it('should return a not found response if the small catch is not found', async () => {
      SmallCatch.findOne.mockResolvedValueOnce(null)
      const h = getMockResponseToolkit()

      const result = await getSmallCatchHandler(
        getSmallCatchRequest('nonexistent-id'),
        h
      )

      expect(result).toBe(NOT_FOUND_SYMBOL)
    })

    it('should call handleServerError if an error occurs while fetching the small catch', async () => {
      const error = new Error('Database error')
      SmallCatch.findOne.mockRejectedValueOnce(error)
      const h = getMockResponseToolkit()

      await getSmallCatchHandler(getSmallCatchRequest('1'), h)

      expect(handleServerError).toHaveBeenCalledWith(
        'Error fetching small catch by ID',
        error,
        h
      )
    })

    it('should return an error response if an error occurs while fetching the the small catch', async () => {
      const error = new Error('Database error')
      SmallCatch.findOne.mockRejectedValueOnce(error)
      const h = getMockResponseToolkit()

      const result = await getSmallCatchHandler(getSmallCatchRequest('1'), h)

      expect(result).toBe(SERVER_ERROR_SYMBOL)
    })
  })

  describe('DELETE /smallCatches/{smallCatchId}', () => {
    const getTransaction = () => ({ commit: jest.fn(), rollback: jest.fn() })

    const getDeleteRequest = (smallCatchId) =>
      getServerDetails({
        params: {
          smallCatchId
        }
      })

    const setUpMocks = ({
      transaction,
      smallCatchCountToDelete = 1,
      smallCatchToDelete = 1
    } = {}) => {
      sequelize.transaction.mockResolvedValueOnce(transaction)
      SmallCatchCount.destroy.mockResolvedValueOnce(smallCatchCountToDelete)
      SmallCatch.destroy.mockResolvedValueOnce(smallCatchToDelete)
    }

    afterEach(() => {
      jest.clearAllMocks()
    })

    it('should return a 204 status code if the small catch is deleted successfully', async () => {
      setUpMocks({ transaction: getTransaction() })

      const result = await deleteSmallCatchHandler(
        getDeleteRequest('2'),
        getMockResponseToolkit()
      )

      expect(result.statusCode).toBe(204)
    })

    it('should return an empty response body when the small catch is deleted successfully', async () => {
      setUpMocks({ transaction: getTransaction() })

      const result = await deleteSmallCatchHandler(
        getDeleteRequest('2'),
        getMockResponseToolkit()
      )

      expect(result.payload).toBeUndefined()
    })

    it('should delete all SmallCatchCount records', async () => {
      const smallCatchId = '2'
      const transaction = getTransaction()
      setUpMocks({ transaction })

      await deleteSmallCatchHandler(
        getDeleteRequest(smallCatchId),
        getMockResponseToolkit()
      )

      expect(SmallCatchCount.destroy).toHaveBeenCalledWith({
        where: { small_catch_id: smallCatchId },
        transaction
      })
    })

    it('should delete all SmallCatch records', async () => {
      const smallCatchId = '2'
      const transaction = getTransaction()
      setUpMocks({ transaction })

      await deleteSmallCatchHandler(
        getDeleteRequest(smallCatchId),
        getMockResponseToolkit()
      )

      expect(SmallCatch.destroy).toHaveBeenCalledWith({
        where: { id: smallCatchId },
        transaction
      })
    })

    it('should commit the transaction on successful deletion', async () => {
      const transaction = getTransaction()
      setUpMocks({ transaction })

      await deleteSmallCatchHandler(
        getDeleteRequest('1'),
        getMockResponseToolkit()
      )

      expect(transaction.commit).toHaveBeenCalled()
    })

    it('should call handleNotFound if no small catch is found to delete', async () => {
      setUpMocks({
        transaction: getTransaction(),
        smallCatchCountToDelete: 1,
        smallCatchToDelete: 0
      })
      const h = getMockResponseToolkit()

      const result = await deleteSmallCatchHandler(getDeleteRequest('0'), h)

      expect(handleNotFound).toHaveBeenCalledWith(
        'Small catch not found for ID: 0',
        h
      )
      expect(result).toBe(NOT_FOUND_SYMBOL)
    })

    it('should rollback the transaction if no small catch is found to delete', async () => {
      const transaction = getTransaction()
      setUpMocks({
        transaction,
        smallCatchCountToDelete: 1,
        smallCatchToDelete: 0
      })
      const h = getMockResponseToolkit()

      await deleteSmallCatchHandler(getDeleteRequest('0'), h)

      expect(transaction.rollback).toHaveBeenCalled()
    })

    it('should call handleServerError if an error occurs while deleting the small catch', async () => {
      sequelize.transaction.mockResolvedValueOnce(getTransaction())
      SmallCatchCount.destroy.mockResolvedValueOnce(0)
      const error = new Error('Database error')
      SmallCatch.destroy.mockRejectedValueOnce(error)
      const h = getMockResponseToolkit()

      const result = await deleteSmallCatchHandler(getDeleteRequest('2'), h)

      expect(handleServerError).toHaveBeenCalledWith(
        'Error deleting small catch',
        error,
        h
      )
      expect(result).toBe(SERVER_ERROR_SYMBOL)
    })

    it('should call handleServerError if an error occurs while deleting the small catch count', async () => {
      sequelize.transaction.mockResolvedValueOnce(getTransaction())
      const error = new Error('Database error')
      SmallCatchCount.destroy.mockRejectedValueOnce(error)
      const h = getMockResponseToolkit()

      const result = await deleteSmallCatchHandler(getDeleteRequest('2'), h)

      expect(handleServerError).toHaveBeenCalledWith(
        'Error deleting small catch',
        error,
        h
      )
      expect(result).toBe(SERVER_ERROR_SYMBOL)
    })

    it('should return an error response if an error occurs while deleting the small catch', async () => {
      sequelize.transaction.mockResolvedValueOnce(getTransaction())
      const error = new Error('Database error')
      SmallCatchCount.destroy.mockResolvedValueOnce(0)
      SmallCatch.destroy.mockRejectedValueOnce(error)

      const result = await deleteSmallCatchHandler(
        getDeleteRequest('2'),
        getMockResponseToolkit()
      )

      expect(result).toBe(SERVER_ERROR_SYMBOL)
    })
  })

  describe('PATCH /smallCatches/{smallCatchId}', () => {
    const getTransaction = () => ({ commit: jest.fn(), rollback: jest.fn() })

    const getSmallCatchRequest = (payload) => ({
      ...getServerDetails(),
      params: { smallCatchId: '1' },
      payload
    })

    const setUpMocks = ({
      transaction,
      foundSmallCatch,
      countsToDestroy = 1
    } = {}) => {
      sequelize.transaction.mockResolvedValueOnce(transaction)
      SmallCatch.findOne.mockResolvedValueOnce(foundSmallCatch)
      SmallCatchCount.destroy.mockResolvedValueOnce(countsToDestroy)
      SmallCatchCount.bulkCreate.mockResolvedValueOnce(0)
    }

    afterEach(() => {
      jest.clearAllMocks()
    })

    const getFullSmallCatchPayload = () => ({
      month: 'DECEMBER',
      released: '2',
      counts: [
        {
          method: 'methods/1',
          count: '2'
        },
        {
          method: 'methods/2',
          count: '2'
        }
      ],
      noMonthRecorded: true
    })

    const getFoundSmallCatch = () => ({
      update: jest.fn().mockResolvedValueOnce({
        toJSON: jest.fn().mockReturnValueOnce({
          id: '33551',
          month: 12,
          released: 2,
          reportingExclude: false,
          noMonthRecorded: true,
          createdAt: '2024-12-11T10:07:11.726Z',
          updatedAt: '2025-01-22T10:55:20.512Z',
          version: '2025-01-22T10:55:20.511Z',
          ActivityId: '133750',
          activity_id: '133750',
          counts: [
            {
              small_catch_id: '33551',
              method_id: '1',
              count: 2,
              SmallCatchId: '33551'
            },
            {
              small_catch_id: '33551',
              method_id: '2',
              count: 2,
              SmallCatchId: '33551'
            }
          ]
        })
      })
    })

    it('should return a 200 status code if the small catch is updated successfully', async () => {
      const transaction = getTransaction()
      const foundSmallCatch = getFoundSmallCatch()
      setUpMocks({ transaction, foundSmallCatch })

      const result = await patchSmallCatchHandler(
        getSmallCatchRequest(getFullSmallCatchPayload()),
        getMockResponseToolkit()
      )

      expect(result.statusCode).toBe(200)
    })

    it('should return the updated small catch in the response', async () => {
      const transaction = getTransaction()
      const foundSmallCatch = getFoundSmallCatch()
      setUpMocks({ transaction, foundSmallCatch })

      const result = await patchSmallCatchHandler(
        getSmallCatchRequest(getFullSmallCatchPayload()),
        getMockResponseToolkit()
      )

      expect(result.payload).toMatchSnapshot()
    })

    it('should delete existing counts if new counts are provided', async () => {
      const transaction = getTransaction()
      const foundSmallCatch = getFoundSmallCatch()
      setUpMocks({ transaction, foundSmallCatch })

      await patchSmallCatchHandler(
        getSmallCatchRequest(getFullSmallCatchPayload()),
        getMockResponseToolkit()
      )

      expect(SmallCatchCount.destroy).toHaveBeenCalledWith({
        where: { small_catch_id: '1' },
        transaction
      })
    })

    it('should create new small catch counts if new counts are provided', async () => {
      const transaction = getTransaction()
      const foundSmallCatch = getFoundSmallCatch()
      setUpMocks({ transaction, foundSmallCatch })

      await patchSmallCatchHandler(
        getSmallCatchRequest(getFullSmallCatchPayload()),
        getMockResponseToolkit()
      )

      expect(SmallCatchCount.bulkCreate).toHaveBeenCalledWith(
        [
          { count: '2', method_id: '1', small_catch_id: '1' },
          { count: '2', method_id: '2', small_catch_id: '1' }
        ],
        { transaction }
      )
    })

    it('should call handleNotFound if the small catch does not exist', async () => {
      const transaction = getTransaction()
      sequelize.transaction.mockResolvedValueOnce(transaction)
      SmallCatch.findOne.mockResolvedValueOnce(undefined)

      const result = await patchSmallCatchHandler(
        getSmallCatchRequest(getFullSmallCatchPayload()),
        getMockResponseToolkit()
      )

      expect(result).toBe(NOT_FOUND_SYMBOL)
    })

    it('should rollback the transaction if deleting small catch count fails', async () => {
      const error = new Error('yesDatabase error')
      const transaction = getTransaction()
      const foundSmallCatch = getFoundSmallCatch()
      sequelize.transaction.mockResolvedValueOnce(transaction)
      SmallCatch.findOne.mockResolvedValueOnce(foundSmallCatch)
      SmallCatchCount.destroy.mockRejectedValueOnce(error)

      await patchSmallCatchHandler(
        getSmallCatchRequest(getFullSmallCatchPayload()),
        getMockResponseToolkit()
      )

      expect(transaction.rollback).toHaveBeenCalled()
    })

    it('should call handleServerError if an error occurs while updating the small catch', async () => {
      const error = new Error('Database error')
      const transaction = getTransaction()
      sequelize.transaction.mockResolvedValueOnce(transaction)
      SmallCatch.findOne.mockRejectedValueOnce({
        update: jest.fn().mockRejectedValueOnce(error)
      })

      const result = await patchSmallCatchHandler(
        getSmallCatchRequest(getFullSmallCatchPayload()),
        getMockResponseToolkit()
      )

      expect(result).toBe(SERVER_ERROR_SYMBOL)
    })

    it('should rollback the transaction if an error occurs while updating the small catch', async () => {
      const error = new Error('Database error')
      const transaction = getTransaction()
      sequelize.transaction.mockResolvedValueOnce(transaction)
      SmallCatch.findOne.mockRejectedValueOnce({
        update: jest.fn().mockRejectedValueOnce(error)
      })

      await patchSmallCatchHandler(
        getSmallCatchRequest(getFullSmallCatchPayload()),
        getMockResponseToolkit()
      )

      expect(transaction.rollback).toHaveBeenCalled()
    })

    it('should rollback the transaction if an error occurs while creating the new small catch counts', async () => {
      const error = new Error('Database error')
      const transaction = getTransaction()
      sequelize.transaction.mockResolvedValueOnce(transaction)
      SmallCatch.findOne.mockResolvedValueOnce(getFoundSmallCatch())
      SmallCatchCount.destroy.mockResolvedValueOnce(0)
      SmallCatchCount.bulkCreate.mockRejectedValueOnce(error)

      await patchSmallCatchHandler(
        getSmallCatchRequest(getFullSmallCatchPayload()),
        getMockResponseToolkit()
      )

      expect(transaction.rollback).toHaveBeenCalled()
    })
  })

  // TODO add table test
})
