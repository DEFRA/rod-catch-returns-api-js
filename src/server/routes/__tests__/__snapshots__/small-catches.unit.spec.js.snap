// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`smallCatches.unit GET /smallCatches/{smallCatchId} should return a 200 status code and the small catch if it is found 1`] = `
{
  "_links": {
    "activity": {
      "href": "http://localhost:3000/api/smallCatches/33551/activity",
    },
    "activityEntity": {
      "href": "http://localhost:3000/api/activities/133750",
    },
    "self": {
      "href": "http://localhost:3000/api/smallCatches/33551",
    },
    "smallCatch": {
      "href": "http://localhost:3000/api/smallCatches/33551",
    },
  },
  "counts": [
    {
      "_links": {
        "method": {
          "href": "http://localhost:3000/api/methods/1",
        },
      },
      "count": 2,
    },
    {
      "_links": {
        "method": {
          "href": "http://localhost:3000/api/methods/3",
        },
      },
      "count": 2,
    },
  ],
  "createdAt": "2024-12-11T10:07:11.726Z",
  "id": "33551",
  "month": "JANUARY",
  "noMonthRecorded": false,
  "released": 1,
  "reportingExclude": false,
  "updatedAt": "2024-12-11T10:07:11.726Z",
  "version": "2024-12-11T10:07:11.729Z",
}
`;

exports[`smallCatches.unit GET /smallCatches/{smallCatchId}/activity should return a 200 status code and the activity if the small catch and activity is found 1`] = `
{
  "_links": {
    "activity": {
      "href": "http://localhost:3000/api/activities/3",
    },
    "catches": {
      "href": "http://localhost:3000/api/activities/3/catches",
    },
    "river": {
      "href": "http://localhost:3000/api/activities/3/river",
    },
    "self": {
      "href": "http://localhost:3000/api/activities/3",
    },
    "smallCatches": {
      "href": "http://localhost:3000/api/activities/3/smallCatches",
    },
    "submission": {
      "href": "http://localhost:3000/api/activities/3/submission",
    },
  },
  "createdAt": "2024-10-10T13:26:30.344Z",
  "daysFishedOther": 0,
  "daysFishedWithMandatoryRelease": 1,
  "id": "3",
  "updatedAt": "2024-10-10T13:26:30.344Z",
  "version": "2024-10-10T13:26:30.345Z",
}
`;

exports[`smallCatches.unit POST /smallCatches should return the created small catch if the call to create it is successful 1`] = `
{
  "_links": {
    "activity": {
      "href": "http://localhost:3000/api/smallCatches/1/activity",
    },
    "activityEntity": {
      "href": "http://localhost:3000/api/activities/3",
    },
    "self": {
      "href": "http://localhost:3000/api/smallCatches/1",
    },
    "smallCatch": {
      "href": "http://localhost:3000/api/smallCatches/1",
    },
  },
  "counts": [
    {
      "_links": {
        "method": {
          "href": "http://localhost:3000/api/methods/1",
        },
      },
      "count": 3,
    },
    {
      "_links": {
        "method": {
          "href": "http://localhost:3000/api/methods/2",
        },
      },
      "count": 2,
    },
    {
      "_links": {
        "method": {
          "href": "http://localhost:3000/api/methods/3",
        },
      },
      "count": 1,
    },
  ],
  "createdAt": "2024-10-11T09:30:57.463+0000",
  "id": "1",
  "month": "FEBRUARY",
  "noMonthRecorded": false,
  "released": 3,
  "reportingExclude": false,
  "updatedAt": "2024-10-11T09:30:57.463+0000",
  "version": "2024-10-11T09:30:57.463+0000",
}
`;
